### ISSUE TITLE
Questions regarding to restfulness of the REST API

### ISSUE THREAD

**Issue by mingchan** [intelsdi-x/snap#633]
***Thanks for sharing this great tool.***
I’m very interested in this tool and want to participate. Since I'm still new to Go, I start with the REST API.  I’ve a few questions and comments on the current version of the rest API. Please take my questions as newbie enquiry and you may have consider some if not all of the comments below already.  I hope that clarifying some of the items may help the community to grow our knowledge base.

* Q1) Task API: What is the difference between “started” and “enabled” task?

* Q2) Task API: What does ‘watch’ task API do?  Wont it be  similar to GET task?
    
Will “tasks/watching” better suite as there should be no verb (watch) used in the URL.

```
GET /v1/tasks/watching/:id/
```

This is assuming that there is no need to put any tasks into the watching bucket before server responding to "tasks/watching."
Is this assumption correct?  If not, a PUT operation may be needed before call GET.

* Q3) Task API: Start and stop running tasks.
Verb (start/stop) should not be used in URL and it may be more intuitive to use the following URL instead.

***start a task***
```
PUT /v1/tasks/running/:id
```

***stop a task***
```
DELETE  /v1/tasks/running/:id 
```

* Q4) Tribe API: Plural noun
It is a good practice to being consistent with using either plural or singular nouns in the URL.
Since the rest nouns are plural like “plugins”, “metrics”, “tasks”, it is more consistent to use “tribes” instead.
```
/v1/tribes/[...]
```

* Q5) Tribe API: add and remove node to and from an agreement. Verb like ‘join/leave” should not be used in URL. May consider the following URLs.

***Join a member node into an agreement given the agreement name***
```
PUT  /v1/tribes/agreements/:agreement_name/members/:node_name
```

***Remove a member node from an agreement given the agreement name***
```
DELETE  /v1/tribes/agreements/:agreement_name/members/:node_name
```

* Q6) Tribe API: Should use plural noun members 

***List tribe member information given the node name***
```
GET /v1/tribes/members/:node_name
```
***List all  members information***
```
GET /v1/tribes/members
```
===============
Thanks,
ming

**Comment by jcooklin** [comment]
Hi @mingchan, thanks for your interest.  

*Q1) Task API: What is the difference between “started” and “enabled” task?*

If a task is disabled it must be enabled before it can be restarted.  Consecutive errors when running a task will cause it to be disabled.

*Q2) Task API: What does ‘watch’ task API do? Wont it be similar to GET task?*

The watch endpoint receives server-sent events.  [This](https://github.com/jcooklin/snap-plugin-collector-cpuid) repo demonstrates creating a snap collector plugin using a stripped down example.  The readme includes an animated gif that demonstrates loading plugins, starting and watching a task.

*Q3) Task API: Start and stop running tasks*

If you use DELETE to "stop" the task how would you express **removing** the task?

*Q4) Tribe API: Plural noun*

In this case *tribe* does not represent a collection so we thought it was a better to go with the singular form.  Under tribe are agreements and members.  
```
s.r.GET("/v1/tribe/agreements", s.getAgreements)
		s.r.POST("/v1/tribe/agreements", s.addAgreement)
		s.r.GET("/v1/tribe/agreements/:name", s.getAgreement)
		s.r.DELETE("/v1/tribe/agreements/:name", s.deleteAgreement)
		s.r.PUT("/v1/tribe/agreements/:name/join", s.joinAgreement)
		s.r.DELETE("/v1/tribe/agreements/:name/leave", s.leaveAgreement)
		s.r.GET("/v1/tribe/members", s.getMembers)
		s.r.GET("/v1/tribe/member/:name", s.getMember)
```

*Q5) Tribe API: add and remove node to and from an agreement. Verb like ‘join/leave” should not be used in URL*

Similar to Q3 above.  DELETE is used to delete the agreement with the same URL.  

*Q6) Tribe API: Should use plural noun members*

I agree.  We currently have the route `"/v1/tribe/member/:name"` and here `member` should be `members`.  I created another issue to capture [this](https://github.com/intelsdi-x/snap/issues/634).

Thanks for your feedback.

**Comment by lynxbat** [comment]
"Tribe" will need to move to "tribes" in the next revision. So might as well capture now to change.

**Comment by jcooklin** [comment]
@lynxbat: Agreed, I'll capture that in the spec for Tribe 2.0 spec which will be posted later this week.

**Comment by mingchan** [comment]
@jcooklin  and @lynxbat  

*Q1) Task API: What is the difference between “started” and “enabled” task?*
>If a task is disabled it must be enabled before it can be restarted. Consecutive errors when running a task will cause it to be disabled.

Here is how I understand a task's *state* transition. Even I use state-like flow, I don't imply that it is implemented in this way. It is more of the notion of having task's state transition as reading through the API.
* {API request to start a task, can I assume the task in stopped state?} --> [Started]
* [Started] --{API request to stop a running task} -->  [Stopped]
* [Started] ---{consecutive errors when running} --> [Disabled]
* [Disabled] ---{API request to enable a  task} --->  [Enabled]
* [Enabled] ---{API request to start a task} --->  [Started]

The question will become what is the benefits of having ‘disabled and enabled” state? What problems it help address?  Without knowing the details, won’t the operation flow be much simpler if we can eliminate the  two extra states, like in below, 

* [Stopped] -- {API request to start a task} --> [Started]
* [Started] -- {API request to stop a running task} -->  [Stopped]
* [Started] -- {consecutive errors when running} --> [Stopped]

*Q2) Task API: What does ‘watch’ task API do? Wont it be similar to GET task?*
>The watch endpoint receives server-sent events. This repo demonstrates creating a snap collector plugin using a stripped down example. The readme includes an animated gif that demonstrates loading plugins, starting and watching a task.

Thanks for the pointer. But that is a command line interface not REST API interface.  Rest API is riding on top of HTTP, and there isn't a streaming of events back to the client, or you have done something differently in here?. HTTP is request and response communication, which implies that client will need to pull the server for events.

There are two cases in here that I can see: use-case-1) Simple best effort to retrieve last x seconds of event log; use-case-2) Getting all events since last pulling to the client. 

Also, seems like using **"watch"** for managing log events is not as correct then; perhaps it is more appropriate to use “/tasks/eventlogs” space for this purpose?

Here is my logic on how to manage the log events. Please let me know if that makes sense or not.

*For use-case-1), simply issuing*
```
GET /tasks/eventlogs/:id
```

It will do for a predefined time as how far back in time the log events are needed . Or Having an additional query param to express the desire amount of time back in the log.  Like,
```
GET /tasks/eventlogs/:id/?duration=5
```
It retrieves log events starting from 5 seconds back till now.


*For use-case-2), it can be done by tracking the last  “GET /tasks/eventlogs/:id” request and returning all log events since then.*

But if keeping and tracking them are resources intensive, we may want to enforce the notion that only tasks specifically put under watch can be watched in this manner.

Then before client can do a **GET** operation, there is a need to put the task under “/task/eventlogs” space by PUT and removing it using DELETE when the 'watching' is no longer needed.

**To start watch the logging events**
```
PUT /tasks/eventlogs/:id
```

***To stop watch the logging events***
```
DELETE /tasks/eventlogs/:id
```
**To retrieve the logging events***
```
GET /tasks/eventlogs/:id
```


*Q3) Task API: Start and stop running tasks*
>If you use DELETE to "stop" the task how would you express removing the task?

They are under different ***spaces*** - “/tasks” and “/tasks/running”

For example, using DELETE on “/tasks/running” space to stop task execution is not the same as removing a task under  “/tasks”.

**“/tasks” space** is used to manage the lifecycle of a task, meaning creation and removal of a task.
**Task creation**
```
POST /tasks
```
**Task removal**
```
DELETE /tasks/:id
```

**“/tasks/running” space** (which is a subspace under “/tasks”) is to manage the running of tasks.

Putting a task-id into “/tasks/running” space is to “start” the task and deleting a task-id from “ “/tasks/running” space is to stop the task.

```
PUT /tasks/running/:id
DELETE /tasks/running/:id
```

*Q5) Tribe API: add and remove node to and from an agreement. Verb like ‘join/leave” should not be used in URL*
>Similar to Q3 above. DELETE is used to delete the agreement with the same URL.

Agreements and their members management are separated and under different spaces.

Under “/tribes/agreements” space, it is for managing the creation/deletion of agreements.

Like “POST /tribes/agreements” and “DELETE /tribes/agreements/:agreement_name” to create and remove an agreement respectively.

The “/tribes/agreements/:agreement_name/members” space is for managing  membership A.K.A. joining/leaving of an agreement.


=======================
Thanks,

Ming

**Comment by lynxbat** [comment]
> The question will become what is the benefits of having ‘disabled and enabled” state? What problems it help address? Without knowing the details, won’t the operation flow be much simpler if we can eliminate the two extra states, like in below,

We have use cases than spawned the creation of these states. But these are based on operational concerns on a large scale not REST API (which represents the CRUD ops but does not define them).
A short example:
* started - a running task
* stopped - a task that is not running
* disabled - a task in a state not allowed to start
* enabled - a task in a state where starting is allowed but is stopped

Disabled has very solid use cases and won't be removed. And argument for "enabled" just really being "stopped" could be made.

> Thanks for the pointer. But that is a command line interface not REST API interface. Rest API is riding on top of HTTP, and there isn't a streaming of events back to the client, or you have done something differently in here?. HTTP is request and response communication, which implies that client will need to pull the server for events.

You are incorrect here. We may need to better explain this feature. It is an event stream sent over a long running HTTP connection. Very similar to the example of SSE on REST in this link(https://jersey.java.net/documentation/latest/sse.html).
Here is an excerpt:
```
SSE is similar to the long-polling mechanism, except it does not send only one message per connection. The client sends a request and server holds a connection until a new message is ready, then it sends the message back to the client while still keeping the connection open so that it can be used for another message once it becomes available. Once a new message is ready, it is sent back to the client on the same initial connection. Client processes the messages sent back from the server individually without closing the connection after processing each message.
```

To call out nitpick, `But that is a command line interface` is not correct. The CLI calls the REST API completely. There is no CLI interface on the daemon other than starting the service.

We don't want a polling mechanism here, especially to try and match a polling GET model for REST. We find a streaming resource on REST as not breaking principles (as seen in the SSE link above). Also, from a functional perspective a streaming event stream that is dynamically "tapped" into is a huge operational feature. It will not be changed to meet a specific RESTFul operation that is non-performant.

The use of the `watch`represents a specific resource on the task that streams the events. It could be renamed to be less confusing (and less command-like) but it will be SSE.

> Tasks and execution are under different spaces - “/tasks” and “/tasks/running”

This feels very unRESTful to me. Tasks are a resource. `running` is a query or tasks of a specific state not a resource.

The tribe API is going to update and we have to reexamine the resource model as a part of that. We can incorporate this into discussion.

Some things that I am recognizing here:
1. We need more documentation on items above that were not obvious to you and reveal gaps in our documentation.
2. We need a good guideline doc for our REST API
3. We should perhaps break up some of your pieces above into individual discussions to better sped time on each one. If you are willing we can do a Google Hangouts.

**Comment by geauxvirtual** [comment]
One correction. There is no state of "enabled" for a task. Enable is used in the REST API to change a task's state from disabled to stopped.

**Comment by lynxbat** [comment]
My mistake.

In this case we won't remove a state change we have in our object model (because of a use case) to make it more RESTful. We could change the way that state is changed of course.

**Comment by lynxbat** [comment]
@mingchan We still have a few items above that we need to nail down as things to iterate on. I Think what is left is 1) `watch` is perhaps a bad word for what we provide 2) `tribe` should be `tribes` 3) `member` should be `members` (tracked at #634) 4) We should reexamine the running commands against tasks as a path. One example of your recommendation would be the Github API using (following) on users (https://developer.github.com/v3/users/followers/).

Did I capture it correctly?

**Comment by mingchan** [comment]
@lynxbat 
Thanks for helping with newbie questions. I agree that with some more documentation and guideline on doc, which will go a long way helping newcomers to quickly understand the interface.  Also thanks for the offer on offline discussions. I will reach out once I spend some more time digest the thread and learn more on code base.

Another topic in mind is HTTP return-codes usage and the guideline on that. I’m sure that is already in your team’s to-do list:-)

Many thanks!

**Comment by lynxbat** [comment]
@mingchan We love the effort you put into this. We want more :)

**Comment by mjbrender** [comment]
Rereading and loving this conversation @mingchan. I'm going to tag this as a question. Lmk if there's more to explore here or whether I should close it. Going to reference it in a documentation issue next.

**Comment by mjbrender** [comment]
Based on this last comment @mingchan, I think we're covered in #749. Closing and thanks again!
